dbt/include/snowflake/macros/materializations/incremental.sql

```diff
@@ -10,10 +10,16 @@
        for faster overall incremental processing.
   #} */
 
-  {% if language == 'sql' and (strategy in ('default', 'append', 'merge') or (unique_key is none)) %}
-    {{ return('view') }}
-  {% else %}  {#--  play it safe -- #}
-    {{ return('table') }}
+  {% if language == "sql" and strategy in ("default", "merge") and merge_tmp_relation_type == "table" %}
```

Comment:
can we simplify and do the `if language == "sql"` condition once?

Reply:
something like this maybe? not tested or anything...
```suggestion
{% if language == "sql"%}
 {% if strategy in ("default", "merge")%}
  {% if merge_tmp_relation_type == "table" %}
    {{ return("table") }}
  {% else %}
    {{ return("view") }}
 {% elif strategy == "append" %}
    {{ return("view") }}
 {% elif unique_key is none %}
    {{ return("view") }}
{% else %}
   {{ return("table") }}
```

Reply:
@colin-rogers-dbt What do you think about alternative pseudo code for doing the `language` condition only once?

It's largely the same as what you proposed. The main difference is that it inverts the `sql` condition so that it doesn't need any nested conditionals.

```
if language != "sql":
    return "table"
else if strategy in ('default', 'merge') and merge_tmp_relation_type == "table":
    return "table"
else if strategy in ('default', 'merge') and merge_tmp_relation_type == "view":
    return "view"
else if strategy in ('default', 'merge', 'append'):
    return "view"
else if unique_key is none:
    return "view"
else:
    return "table"
```

Reply:
The != logic is definitely much cleaner

Reply:
if `merge_tmp_relation_type == "table"` would we ever return "view"?

Reply:
In that case I think we can actually simplify this further:
```
if language != "sql":
    return "table"
else if strategy in ('append'):
    return "view"
else if merge_tmp_relation_type == "table":
    return "table"
else if merge_tmp_relation_type == "view":
    return "view"
else if unique_key is none:
    return "view"
else:
    return "table"
```

Reply:
do you think that would make the fact that` merge_tmp_relation_type` only technically applies to the merge/default strategy as they are the same in snowflake?

Reply:
@colin-rogers-dbt Love the way you are trying to make the logic as simple as possible! We just need to make sure to cover each of the moving pieces.

My understanding is that dbt-snowflake supports [three different incremental strategies](https://github.com/dbt-labs/dbt-snowflake/blob/4b7ee7372d87efc2b9c6929203fcfb41930b0b3e/dbt/adapters/snowflake/impl.py#L236-L237):
1. `merge` (which is the `default`)
1. `append`
1. `delete+insert`

Something to point out: there's the largely implicit case when `strategy == "delete+insert"` to consider. In the current version of the PR, it comes into play when all other strategies have already been considered _and_ `unique_key is none`. If your suggestion is implemented as-is, then the configuration of `merge_tmp_relation_type` would affect `merge` and `delete+insert`, but it wouldn't affect `append`.

Leaning into your idea, I suspect it would be like:
```
if language != "sql":
    return "table"
else if strategy in "append":
    return "view"
else if strategy == "delete+insert" and unique_key is none:
    return "view"
else if merge_tmp_relation_type == "table":
    return "table"
else if merge_tmp_relation_type == "view":
    return "view"
else:
    return "table"
```

Or a more future-proof version presuming some other strategy might be added at a later date:
```
if language != "sql":
    return "table"
else if strategy in "append":
    return "view"
else if strategy == "delete+insert" and unique_key is none:
    return "view"
else if strategy in ("default", "merge") and merge_tmp_relation_type == "table":
    return "table"
else if strategy in ("default", "merge") and merge_tmp_relation_type == "view":
    return "view"
else:
    return "table"
```

Side note: [here](https://github.com/dbt-labs/dbt-snowflake/issues/323#issuecomment-1382015544) is a proposal to make both the logic and the documentation as simple as possible. It would need buy-in from @Fleid. That proposal is non-blocking for this PR though.